### PR TITLE
Mismatch of inferred type and return type note

### LINKED ISSUE #2529: Better message when inferred variable type conflicts with later use
If you can come up with a better name for this issue, please let me know, this one is poor. The naming difficulty also lead to difficulty searching to see if this already existed, so apologies in advance if this is a duplicate issue.

When running this code through the typechecker:
```python
from typing import Union, Dict, List
def g() -> List[Union[str, int]]:
    x = ['a']
    x += [1]
    return x

def f() -> Dict[str, Union[str, int]]:
    d = {'a': 'a'}
    d['b'] = 1
    return d
```

I get this output:
```
test.py: note: In function "g":
test.py:4: error: List item 0 has incompatible type "int"
test.py:5: error: Incompatible return value type (got List[str], expected List[Union[str, int]])
test.py: note: In function "f":
test.py:9: error: Incompatible types in assignment (expression has type "int", target has type "str")
test.py:10: error: Incompatible return value type (got Dict[str, str], expected Dict[str, Union[str, int]])
```

With both these functions, mypy is assuming the type on assignment, then having type errors due to conflicts with the explicitly declared return type.

Surely there should be a way to correctly infer the original type based on the return type?

### PR DESCRIPTION
**Pull request by quartox**
Fixes #2529 

Not sure about this function sitting in `checker.py`, might need to move it to `messages.py` but didn't look into that yet.

I use the `context.expr.node.is_inferred` to determine if the type of the variable is inferred to determine if the message is printed. Apparently `is_inferred` is still `True` even when the adding a type comment in the code. See below:

```
def i() -> List[Union[int, float]]:
    x = [1] # Type: List[int]
    return x
```

Specifically, the `context.expr.node.is_inferred=True` for `x` when I would expect it to be `False`.

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -3115,3 +3120,23 @@ def push_class(self, info: TypeInfo) -> Iterator[None]:
...
+    if (msg.startswith(messages.INCOMPATIBLE_RETURN_VALUE_TYPE) and
+      isinstance(subtype, Instance) and isinstance(supertype, Instance) and
+      subtype.type.fullname() == supertype.type.fullname() and
+      subtype.args and supertype.args and
+      isinstance(context, ReturnStmt) and isinstance(context.expr, NameExpr) and
+      isinstance(context.expr.node, Var) and context.expr.node.is_inferred):
+        for subtype_arg, supertype_arg in zip(subtype.args, supertype.args):
+            if not is_subtype(subtype_arg, supertype_arg):
+                return ''
+        return 'Perhaps a type annotation is missing here? Suggestion: ' + supertype_str
```
Comment (JukkaL) [inline mypy/checker.py]: Might be better to also mention the name of the variable to make it clear that the proposed annotation is for the variable, not for the return statement. For example, `Perhaps you need a type annotation for "x"? Suggestion: List[float]`

**Comment by quartox** [comment]
Made all of the fixes but moving it to messages.py

Since I use the `is_subtype` function from the subtypes module, moving this to messages.py creates a circular import between messages.py and subtypes.py that apparently breaks the imports.

I will try to figure out how to fix this, but might need some ideas.

**Comment by ilevkivskyi** [comment]
> moving this to messages.py creates a circular import between messages.py and subtypes.py that apparently breaks the imports.

If you really need this, then you maybe can import ``subtypes`` inside your function. Have you tried this?

**Comment by quartox** [comment]
Thanks, that did it.

File: test-data/unit/check-functions.test [hunk 2]
```diff
@@ -2067,3 +2067,26 @@ def some_method(self: badtype): pass # E: Name 'badtype' is not defined
...
+# N: Perhaps you need a type annotation for "x"? Suggestion: Dict[str, Union[int, str]]
+
+def h() -> List[Union[int, float]]:
+    x = ['a']
+    return x # E: Incompatible return value type (got List[str], expected List[Union[int, float]])
+
+def i() -> List[Union[int, float]]:
+    x = [1] # Type: List[int]
+    return x # E: Incompatible return value type (got List[int], expected List[Union[int, float]]) \
+# N: Perhaps you need a type annotation for "x"? Suggestion: List[Union[int, float]]
```
Comment (JukkaL) [inline test-data/unit/check-functions.test]: Almost there! I'd suggest adding test cases that cover cases that are mostly like these but that don't actually trigger the new message (i.e. one of the conditions in `make_inferred_type_note` is false).

Comment (JukkaL) [inline test-data/unit/check-functions.test]: One more thing: Can you split the test case into multiple separate test cases? It's getting pretty big now. At least cases where we generate the note and where we don't generate a note could live in separate test cases.

**Comment by quartox** [comment]
Added more test cases.

**Pull request merged**